### ISSUE TITLE
isXMLDoc() - Uncaught TypeError: Cannot read property 'namespaceURI' of undefined

### ISSUE THREAD

**Issue by borzaka** [jquery/jquery#4782]
### Description ###

`isXMLDoc()` throws an error if `undefined` given, starting from version 3.4.0

#### Test script:
`jQuery.isXMLDoc(undefined)`

#### jQuery 3.3.1 result:
`false`

#### jQuery 3.5.1 (currently the latest, but starting from 3.4.0) result:
`Uncaught TypeError: Cannot read property 'namespaceURI' of undefined`

#### Techinfo

Before the **Sizzle** update from 2.3.3 to 2.3.4 (see related commit below), `isXML()` function looked liked this:
```
isXML = Sizzle.isXML = function( elem ) {
	// documentElement is verified for cases where it doesn't yet exist
	// (such as loading iframes in IE - #4833)
	var documentElement = elem && (elem.ownerDocument || elem).documentElement;
	return documentElement ? documentElement.nodeName !== "HTML" : false;
};
```

This **doesn't throw an error**, when the `elem` is `undefined`, because it checks it.

In the update from 3.4.0, it looks like this:
```
isXML = Sizzle.isXML = function( elem ) {
	var namespace = elem.namespaceURI,
		docElem = ( elem.ownerDocument || elem ).documentElement;

	// Support: IE <=8
	// Assume HTML when documentElement doesn't yet exist, such as inside loading iframes
	// https://bugs.jquery.com/ticket/4833
	return !rhtml.test( namespace || docElem && docElem.nodeName || "HTML" );
};
```

This **throws an error** when the `elem` is `undefined`:
`Uncaught TypeError: Cannot read property 'namespaceURI' of undefined`

**Please, if you can, bring back the previous functionality: `undefined` is not an XML node, so it should return a false boolean instead of a TypeError.**

Thanks in advance!

### Additional info ###

Related commit: https://github.com/jquery/jquery/commit/d940bc063cd5ae09bf52067bbe046fda88f1f687 (from line 562)
Related blog-post: https://blog.jquery.com/2019/04/10/jquery-3-4-0-released/

**Comment by dmethvin** [comment]
Can you post a test case showing when this happens? Also can you use the latest jQuery 3 in your test case?

Be aware that version 2.x is no longer supported. If there is a bug to fix we'd only be fixing it in jQuery 3.x, or 4.x if it requires a breaking change.

**Comment by borzaka** [comment]
I'am afraid if I present my use case, the attention will move away from my finding. Which is:
jQuery 3.3.1 returns a different outcome for the same function and argument then 3.5.1

jQuery 3.3.1 returns: `false`
jQuery 3.5.1 returns: `Uncaught TypeError: Cannot read property 'namespaceURI' of undefined`
For this line of code: `jQuery.isXMLDoc(undefined)`

jQuery 3.5.1 doesn't check the argument, 3.3.1 does. `undefined` is not an XML node, so in my interpretation it should return `false`, as version 3.3.1 does.

I'am talking about jQuery 3.x, not 2.x, only 2.x I mentioned is Sizzle update from 2.3.3 to 2.3.4, where this `isXML()` function is changed and causing this new TypeError since 3.4.0. Linked under _Additional info_ in my original post.

**Comment by mgol** [comment]
As indicated in the signature of that method on [its API page](https://api.jquery.com/jQuery.isXMLDoc/), only arguments of Element type are allowed as input to that method. Behavior for all other inputs is undefined as jQuery generally doesn't validate whether input is of a correct type.

If you want to make sure you only use allowed types, using TypeScript with jQuery typings is a good idea.

I'll let it open for discussion for now but generally we've been leaning towards not handling `undefined` inputs in APIs where it's not expected.

**Comment by timmywil** [comment]
We're deliberating whether we want to add back the falsey check.

**Comment by timmywil** [comment]
I think we're willing to add back the falsey check, but we'd all feel better if there was a valid use case presented.

> the attention will move away from my finding

Just the opposite btw. It will grant some validity to your finding.

**Comment by borzaka** [comment]
Sorry for the late reply. Here is a part of that code working with ajax responses.

```
mainComplete = function(xhr, status) {
    var result
    var resultText

    if ($.isXMLDoc(xhr.responseXML) && xhr.responseXML.documentElement.nodeName === 'result') {
        result = processCoreResponse(xhr.responseXML)
    }

    resultText = result ? (result.content[''] || '') : xhr.responseText

    ...
}
```

The `$.isXMLDoc(xhr.responseXML)` is failing because of the lack of the falsey check as before.

Could you reopen this issue please?

**Comment by mgol** [comment]
Thanks for the use case, reopening.

**Comment by mgol** [comment]
PR: #4814

**Comment by mgol** [comment]
PR to jQuery (`master` branch, for jQuery 4.0): #4814
PR to Sizzle (for jQuery 3.6.0): https://github.com/jquery/sizzle/pull/474

**Comment by mgol** [comment]
Landed in jquery/sizzle#474 & #4814 but we still need to update Sizzle on `3.x-stable`.

**Comment by mgol** [comment]
Reopening as it awaits a Sizzle update.

**Comment by mgol** [comment]
Fixed on `3.x-stable` (jQuery 3.x) by #4846.
Previously, it was only fixed on `main` (jQuery 4.0) by #4814.

Closing.
